### ISSUE TITLE
Citeseer data set accuracy

### ISSUE THREAD

**Issue by xuhaiyun42** [tkipf/pygcn#27]
Dear professor,
Hello!
I am very interesting in your recent GCN work.
Thanks for sharing the code, I used the GCN network to run the citeseer database, but the accuracy could not reach 70.3. How did you set the parameters to run so high? Thanks a lot for sharing the code, anyway.

Many thanks for your help.

**Comment by tkipf** [comment]
Note that this repository uses different dataset splits and a slightly different model architecture than in our original paper. For an exact replication of the experiments in our paper, please have a look at this repository: https://github.com/tkipf/gcn

**Comment by xuhaiyun42** [comment]
Thank you for your prompt reply!
I used the data of GCN with tensorflow version to save the features, adjacency matrix and label data set,and put them into pytorch, but the accuracy of tensorflow version could not be reproduced, for citeseer data set

**Comment by tkipf** [comment]
Yes, the model implementation is slightly different for the PyTorch
version. In this version, the adjacency matrix is normalized by
left-multiplication with the inverse of the degree matrix (instead of the
symmetric normalization used in the paper). Additionally, the first layer
doesn’t use Dropout (if I remember correctly) since I didn’t have time to
replicate the sparse dropout implementation from the TensorFlow-GCN
implementation in PyTorch. The performance should be rather similar though.
If you want to get better performance, you can also play around with
different activation functions and regularizers (like LayerNorm or
BatchNorm).

**Comment by xuhaiyun42** [comment]
Thank you again for your reply！ which is of great help to me.

**Comment by xuhaiyun42** [comment]
Dear professor,

Hello!
It makes sense that you load the cora dataset this way and construct the adjacency matrix.  
idx_features_labels = np.genfromtxt("{}{}.content".format(path, dataset), dtype=np.dtype(str))
print(idx_features_labels.shape)
features = sp.csr_matrix(idx_features_labels[:, 1:-1], dtype=np.float32)
labels = encode_onehot(idx_features_labels[:, -1])

# build graph
idx = np.array(idx_features_labels[:, 0], dtype=np.int32)
idx_map = {j: i for i, j in enumerate(idx)}
edges_unordered = np.genfromtxt("{}{}.cites".format(path, dataset), dtype=np.int32)
edges = np.array(list(map(idx_map.get, edges_unordered.flatten())), dtype=np.int32).reshape(edges_unordered.shape)
adj = sp.coo_matrix((np.ones(edges.shape[0]), (edges[:, 0], edges[:, 1])), shape=(labels.shape[0], labels.shape[0]), dtype=np.float32)
How do you load data for Cornell datasets in a WebKB dataset?

The dataset can be seen in the attachment.I hope to get your help. Thank you very much!

**Comment by swyu0711** [comment]
Hello. I also run the pytorch version GCN on citeseer dataset and the accuracy is 69.65%. Furthermore, the accuracy differed everytime on the cora dataset provided by this repository. However, the accuracy is invariable on datasets provided by the original GCN repository https://github.com/tkipf/gcn. 
I can't understand why. I really want to know whether someone else got the same results. 
Thank you.

> Dear professor,
> Hello!
> I am very interesting in your recent GCN work.
> Thanks for sharing the code, I used the GCN network to run the citeseer database, but the accuracy could not reach 70.3. How did you set the parameters to run so high? Thanks a lot for sharing the code, anyway.
> 
> Many thanks for your help.

**Comment by tkipf** [comment]
The dataset splits used in the PyGCN repository are different from the ones
used in our original paper. If you would just like to reproduce the results
of the original GCN paper, then please use the TensorFlow repository :)

There are also some small implementation changes (due to the nature of
PyTorch) that affect results.

**Comment by bapriddy** [comment]
@tkipf 

Dear Thomas, if you have time could you elaborate just a bit on the "implementation changes" in the Pytorch version you mentioned above?  I'm not necessarily interested in the cora data or those results but more interested in training on other graphs/datasets.  Maybe even a parallel version at some point down the road using Horovod or Keras.

Thanks

**Comment by tkipf** [comment]
I think it’s best to compare the two implementations side by side if you’re
interested in the precise differences (both implementations are fairly
simple).

For a more modern and flexible implementation in PyTorch, I recommend
having a look at
https://github.com/dmlc/dgl

**Comment by bapriddy** [comment]
@tkipf 

Thanks Again!!

**Comment by scottjiao** [comment]
One of the most important reason I think is that there is no available api in pytorch by which the dropout on sparse input can be implemented. In GCN-tensorflow-version Dr. Kipf implement sparse dropout by tf.sparse_retain but this api can not be found in torch. Because dropout is an important hyperparameterfor GCN (empirically) thus we may not have the ability to recover the accuracy without solving this problem.



> Hello. I also run the pytorch version GCN on citeseer dataset and the accuracy is 69.65%. Furthermore, the accuracy differed everytime on the cora dataset provided by this repository. However, the accuracy is invariable on datasets provided by the original GCN repository https://github.com/tkipf/gcn.
> I can't understand why. I really want to know whether someone else got the same results.
> Thank you.
> 
> > Dear professor,
> > Hello!
> > I am very interesting in your recent GCN work.
> > Thanks for sharing the code, I used the GCN network to run the citeseer database, but the accuracy could not reach 70.3. How did you set the parameters to run so high? Thanks a lot for sharing the code, anyway.
> > Many thanks for your help.

**Comment by tkipf** [comment]
The dataset splits in this repository are different from the ones used in
the paper / used in the TesnorFlow gcn repository. This repository is not
intended as an exact replication of the setting in our paper, so slight
differences in model performance are to be expected (even beyond the
dataset split detail).

**Comment by Yfhu1103** [comment]
> Dear professor,
> Hello!
> I am very interesting in your recent GCN work.
> Thanks for sharing the code, I used the GCN network to run the citeseer database, but the accuracy could not reach 70.3. How did you set the parameters to run so high? Thanks a lot for sharing the code, anyway.
> 
> Many thanks for your help.

Hello, I want to know where can I download the citeseer dataset in the form similar to cora dataset in this implementation(citeseer.cites, citeseer.content).
Thank you a lot!

**Comment by Riting-Xia** [comment]
> Hello. I also run the pytorch version GCN on citeseer dataset and the accuracy is 69.65%. Furthermore, the accuracy differed everytime on the cora dataset provided by this repository. However, the accuracy is invariable on datasets provided by the original GCN repository https://github.com/tkipf/gcn.
> I can't understand why. I really want to know whether someone else got the same results.
> Thank you.
> 
> > Dear professor,
> > Hello!
> > I am very interesting in your recent GCN work.
> > Thanks for sharing the code, I used the GCN network to run the citeseer database, but the accuracy could not reach 70.3. How did you set the parameters to run so high? Thanks a lot for sharing the code, anyway.
> > Many thanks for your help.

Hello!
Have you found the reason why the accuracy differed everytime on the cora dataset provided by this repository?  If you know the reason, please tell me, thank you.

**Comment by fansariadeh** [comment]
> Note that this repository uses different dataset splits and a slightly different model architecture than in our original paper. For an exact replication of the experiments in our paper, please have a look at this repository: https://github.com/tkipf/gcn

Dear Dr Kipf,

I am a big fan of your works and really interested in this code you shared.
Regarding Citeseer dataset, I have download it from **https://github.com/kimiyoung/planetoid**, which hopefully is the same data you ahve used.
My problem is reading this file and specifying which file creates graph and which one makes edges.
Can you please elaboarte it more.
Thank you in advance.

Cheers,

**Comment by fansariadeh** [comment]
> Dear professor,
> Hello!
> I am very interesting in your recent GCN work.
> Thanks for sharing the code, I used the GCN network to run the citeseer database, but the accuracy could not reach 70.3. How did you set the parameters to run so high? Thanks a lot for sharing the code, anyway.
> 
> Many thanks for your help.

Regarding Citeseer dataset, I have download it from https://github.com/kimiyoung/planetoid, which hopefully is the same data you ahve used.
My problem is reading this file and specifying which file creates graph and which one makes edges.
Can you please elaboarte it more.
Thank you in advance.

**Comment by fansariadeh** [comment]
> Dear professor, Hello! It makes sense that you load the cora dataset this way and construct the adjacency matrix. idx_features_labels = np.genfromtxt("{}{}.content".format(path, dataset), dtype=np.dtype(str)) print(idx_features_labels.shape) features = sp.csr_matrix(idx_features_labels[:, 1:-1], dtype=np.float32) labels = encode_onehot(idx_features_labels[:, -1]) # build graph idx = np.array(idx_features_labels[:, 0], dtype=np.int32) idx_map = {j: i for i, j in enumerate(idx)} edges_unordered = np.genfromtxt("{}{}.cites".format(path, dataset), dtype=np.int32) edges = np.array(list(map(idx_map.get, edges_unordered.flatten())), dtype=np.int32).reshape(edges_unordered.shape) adj = sp.coo_matrix((np.ones(edges.shape[0]), (edges[:, 0], edges[:, 1])), shape=(labels.shape[0], labels.shape[0]), dtype=np.float32) How do you load data for Cornell datasets in a WebKB dataset? The dataset can be seen in the attachment.I hope to get your help. Thank you very much!

Regarding Citeseer dataset, I have download it from https://github.com/kimiyoung/planetoid, which hopefully is the same data you ahve used.
My problem is reading this file and specifying which file creates graph and which one makes edges.
Can you please elaboarte it more.
Thank you in advance.

**Comment by fansariadeh** [comment]
> > Hello. I also run the pytorch version GCN on citeseer dataset and the accuracy is 69.65%. Furthermore, the accuracy differed everytime on the cora dataset provided by this repository. However, the accuracy is invariable on datasets provided by the original GCN repository https://github.com/tkipf/gcn.
> > I can't understand why. I really want to know whether someone else got the same results.
> > Thank you.
> > > Dear professor,
> > > Hello!
> > > I am very interesting in your recent GCN work.
> > > Thanks for sharing the code, I used the GCN network to run the citeseer database, but the accuracy could not reach 70.3. How did you set the parameters to run so high? Thanks a lot for sharing the code, anyway.
> > > Many thanks for your help.
> 
> Hello!
> Have you found the reason why the accuracy differed everytime on the cora dataset provided by this repository? If you know the reason, please tell me, thank you.

Regarding Citeseer dataset, I have download it from https://github.com/kimiyoung/planetoid, which hopefully is the same data you ahve used.
My problem is reading this file and specifying which file creates graph and which one makes edges.
Can you please elaboarte it more.
Thank you in advance.
